tokens-accounting/src/storage/db.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::BTreeMap, ops::Neg};
+
+use accounting::{
+    combine_amount_delta, combine_data_with_delta, DeltaAmountCollection, DeltaDataUndoCollection,
+};
+use common::{chain::tokens::TokenId, primitives::Amount};
+use utils::tap_error_log::LogError;
+
+use crate::{
+    data::{TokensAccountingDeltaData, TokensAccountingDeltaUndoData},
+    error::Error,
+    FlushableTokensAccountingView, TokenData, TokensAccountingView,
+};
+
+use super::{TokensAccountingStorageRead, TokensAccountingStorageWrite};
+
+#[must_use]
+pub struct TokensAccountingDB<S>(S);
+
+impl<S: TokensAccountingStorageRead> TokensAccountingDB<S> {
+    pub fn new(store: S) -> Self {
+        Self(store)
+    }
+}
+
+impl<S: TokensAccountingStorageWrite> TokensAccountingDB<S> {
+    pub fn merge_with_delta(
+        &mut self,
+        other: TokensAccountingDeltaData,
+    ) -> Result<TokensAccountingDeltaUndoData, Error> {
```

Comment:
Can we somehow make sure this happens inside a database transaction? I.e. restricting `S`?

Reply:
Not sure, transaction is a concept not known here. We actually enforce it by implementing `TokensAccountingStorageWrite` only for transaction.

Reply:
> Not sure, transaction is a concept not known here. We actually enforce it by implementing `TokensAccountingStorageWrite` only for transaction.

That sounds good enough, thanks.